src/main/java/io/curity/identityserver/plugin/aws/authentication/CallbackRequestHandler.java

```diff
@@ -38,118 +28,241 @@
+                                _config.getClientSecret()).getBytes()))
+                .method("POST")
+                .response();
+        int statusCode = tokenResponse.statusCode();
 
-    public Map<String, Object> getTokens(String tokenEndpoint, String clientId, String clientSecret, String code, String state) {
+        if (statusCode != 200)
+        {
+            if (_logger.isInfoEnabled())
+            {
+                _logger.info("Got error response from token endpoint: error = {}, {}", statusCode,
+                        tokenResponse.body(HttpResponse.asString()));
+            }
 
-        _oauthClient.validateState(state);
+            throw _exceptionFactory.internalServerException(ErrorCode.EXTERNAL_SERVICE_ERROR);
+        }
 
-        try {
-            ParamBuilder postData = new ParamBuilder();
-            postData.addPair(PARAM_CODE, code)
-                    .addPair(PARAM_GRANT_TYPE, PARAM_GRANT_TYPE_AUTHORIZATION_CODE)
-                    .addPair(SCOPE, AWSAuthenticatorRequestHandler.getScope(_config))
-                    .addPair(PARAM_REDIRECT_URI, _oauthClient.getCallbackUrl());
-            UrlEncodedFormEntity data = new UrlEncodedFormEntity(postData.getPairs());
+        return _json.fromJson(tokenResponse.body(HttpResponse.asString()));
+    }
 
-            HttpPost post = new HttpPost(tokenEndpoint);
-            post.addHeader(HttpHeaders.CONTENT_TYPE, CONTENT_TYPE);
-            post.setHeader(HttpHeaders.AUTHORIZATION, "Basic " + Base64.getEncoder().encodeToString((clientId + ":" + clientSecret).getBytes()));
-            post.setEntity(data);
+    private WebServiceClient getWebServiceClient()
+    {
+        Optional<HttpClient> httpClient = _config.getHttpClient();
 
-            HttpResponse response = _client.execute(post);
-            if (response.getStatusLine().getStatusCode() != HttpStatus.SC_OK) {
-                _logger.debug("Got error response from token endpoint {}", response.getStatusLine());
+        if (httpClient.isPresent())
+        {
+            return _webServiceClientFactory.create(httpClient.get()).withHost(_config.getDomain().getHost());
+        }
+        else
+        {
+            return _webServiceClientFactory.create(URI.create(_config.getDomain().toString()));
+        }
+    }
 
-                throw _exceptionFactory.internalServerException(ErrorCode.INVALID_SERVER_STATE, "INTERNAL SERVER ERROR");
+    private void handleError(CallbackGetRequestModel requestModel)
+    {
+        if (!Objects.isNull(requestModel.getError()))
+        {
+            if ("access_denied".equals(requestModel.getError()))
+            {
+                _logger.debug("Got an error from AWS: {} - {}", requestModel.getError(),
+                        requestModel.getErrorDescription());
+
+                throw _exceptionFactory.redirectException(
+                        _authenticatorInformationProvider.getAuthenticationBaseUri().toASCIIString());
```

Comment:
probably don't want to use `toASCIIString` here. If I name my authentication profile's authenticate endpoint `öööö`, this will break even though it's a valid URL.

Reply:
okay